File: src/main/java/com/study/csdojo/Problem9_NthElement.java
package com.study.csdojo;

public class Problem9_NthElement {

    private static class Node {
        private int value;
        private Node child;

        public Node(int value, Node child) {
            this.value = value;
            this.child = child;
        }

        @Override
        public String toString() {
            return Integer.toString(value);
        }
    }

    public static void main(String[] args) {
        // NOTE: The following input values will be used for testing your solution.
        Node current = new Node(1, null);
        for (int i = 2; i < 8; i++) {
            current = new Node(i, current);
        }
        Node head = current;
        // head = 7 -> 6 -> 5 -> 4 -> 3 -> 2 -> 1 -> (null)

        Node current2 = new Node(4, null);
        for (int i = 3; i > 0; i--) {
            current2 = new Node(i, current2);
        }
        Node head2 = current2;
        // head2 = 1 -> 2 -> 3 -> 4 -> (null)

        System.out.println("should return 1: " + nthFromLast(head, 1));
        System.out.println("should return 5: " + nthFromLast(head, 5));
        System.out.println("should return 3: " + nthFromLast(head2, 2));
        System.out.println("should return 1: " + nthFromLast(head2, 4));
        System.out.println("should return null: " + nthFromLast(head2, 5));
        System.out.println("should return null: " + nthFromLast(null, 1));
    }


    // Implement your function below.
    public static Node nthFromLast(Node head, int n) {
        Node left = head;
        Node right = head;
        for (int i = 0; i < n; i++) {
            if (right == null) {
                return null;
            }
            right = right.child;
        }
        while (right != null) {
            left = left.child;
            right = right.child;
        }
        return left;
    }

    public static Node myNthFromLast(Node head, int n) {
        if (head == null) {
            return null;
        }
        int count = 0;
        Node tmp = head;
        while (tmp != null) {
            count++;
            tmp = tmp.child;
        }
        if (count < n) {
            return null;
        }
        tmp = head;
        for (int i = 0; i < count - n; i++) {
            tmp = tmp.child;
        }
        return tmp;
    }


    //  NOTE: Feel free to use the following function for testing.
    //  It converts the given linked list into an easy-to-read string format.
    //  Example: 7 -> 6 -> 5 -> 4 -> 3 -> 2 -> 1 -> (null)
    public static String linkedListToString(Node head) {
        Node current = head;
        StringBuilder sb = new StringBuilder();
        while (current != null) {
            sb.append(current.value);
            sb.append(" -> ");
            current = current.child;
        }
        sb.append("(null)");
        return sb.toString();
    }

}

File: src/main/java/com/study/csdojo/Problem9_NthElement.py
class Node:

    def __init__(self, value, child = None):
        self.value = value
        self.child = child

    # The string representation of this node. Will be used for testing.
    def __str__(self):
        return str(self.value)


def nth_from_last(head, n):
    left = head
    right = head
    for i in range(n):
        if right is None:
            return None
        right = right.child
    while right:
        left = left.child
        right = right.child
    return left

def my_nth_from_last(head, n):
    if head is None:
        return None
    count = 0
    tmp = head
    while tmp:
        tmp = tmp.child
        count += 1
    if count < n:
        return None
    tmp = head
    for i in range(count - n):
        tmp = tmp.child
    return tmp


def revert(head):
    if head is None:
        return None
    tmp = head
    last = None
    while tmp:
        next = tmp.child
        tmp.child = last
        last = tmp
        tmp = next
    head = last
    return head


def linked_list_to_string(head):
    current = head
    str_list = []
    while current:
        str_list.append(str(current.value))
        current = current.child
    str_list.append('(None)')
    return ' -> '.join(str_list)

if __name__ == '__main__':
    current = Node(1)
    for i in range(2, 8):
        current = Node(i, current)
    head = current
    # head = 7 -> 6 -> 5 -> 4 -> 3 -> 2 -> 1 -> (None)

    current2 = Node(4)
    for i in range(3, 0, -1):
        current2 = Node(i, current2)
    head2 = current2
    # head2 = 1 -> 2 -> 3 -> 4 -> (None)

    print("%s should return 1." % (nth_from_last(head, 1)))
    print("%s should return 5." % (nth_from_last(head, 5)))
    print("%s should return 3." % (nth_from_last(head2, 2)))
    print("%s should return 1." % (nth_from_last(head2, 4)))
    print("%s should return None." % (nth_from_last(head2, 5)))
    print("%s should return None." % (nth_from_last(None, 1)))

    print("Reverted head: %s" % (linked_list_to_string(revert(head))))
    print("Reverted head2: %s" % (linked_list_to_string(revert(head2))))